### PR TITLE
ci: Move Windows MSVC build from AppVeyor to Cirrus

### PR DESCRIPTION
**Pull request by hebasto**
This PR:
- gets rid off unreliable AppVeyor CI
- places all CI configs in one place
- allows to enable functional tests in the future (using a persistent worker)

| no populated `vcpkg` cache | populated `vcpkg` cache |
|---|---|
| Screenshot from 2021-09-02 15-47-04 | Screenshot from 2021-09-02 14-06-26 |

Currently, AppVeyor builds take about 44..48 minutes.

### CODE REVIEW

**Comment by hebasto** [comment]
The warmed up vcpkg cache [reduces](https://cirrus-ci.com/task/6335159239704576) build time by about 40 minutes.

**Comment by MarcoFalke** [comment]
Still takes longer than the appveyor 38 minutes right now?

**Comment by hebasto** [comment]
> Still takes longer than the appveyor 38 minutes right now?

Right.

**Comment by hebasto** [comment]
@MarcoFalke 
> Still takes longer than the appveyor 38 minutes right now?

Does it make sense to move this task into the DrahtBot infrastructure? Or just close it?

**Comment by hebasto** [comment]
> Still takes longer than the appveyor 38 minutes right now?

It takes 27 minutes now.

> The time is spent on installing `msbuild` tools, that are preinstalled on AppVeyor.

Fixed by using `visualstudio2019` image.

**Comment by fanquake** [comment]
What is the status of this? Is the intention to replace Appveyor? If not, do we need two Windows builds? 

An empty PR description is not really useful.

**Comment by MarcoFalke** [comment]
(Also needs appveyor removed)

**Comment by hebasto** [comment]
@sipsorcery 

It seems [`Qt5.12.11_x64_static_vs2019_16101.zip`](https://github.com/sipsorcery/qt_win_binary/releases/download/qt51211x64_static_vs2019_16101/Qt5.12.11_x64_static_vs2019_16101.zip) is not compatible with the `msbuild` in the `cirrusci/windowsservercore:visualstudio2019` image:

```
LINK : fatal error C1900: Il mismatch between 'P1' version '20210202' and 'P2' version '20210113' [C:\Users\ContainerAdministrator\AppData\Local\Temp\cirrus-ci-build\build_msvc\bitcoin-qt\bitcoin-qt.vcxproj]
LINK : fatal error LNK1257: code generation failed [C:\Users\ContainerAdministrator\AppData\Local\Temp\cirrus-ci-build\build_msvc\bitcoin-qt\bitcoin-qt.vcxproj]
LINK : fatal error LNK1327: failure during running link.exe [C:\Users\ContainerAdministrator\AppData\Local\Temp\cirrus-ci-build\build_msvc\bitcoin-qt\bitcoin-qt.vcxproj]
```

So, I switched back to [`Qt5.12.10_x64_static_vs2019_1694.zip`](https://github.com/sipsorcery/qt_win_binary/releases/download/qt51210x64_vs2019_1694/Qt5.12.10_x64_static_vs2019_1694.zip) for now.

May I ask you to build `Qt5.12.11_x64_static_vs2019_1694.zip` -- the recent Qt 5.12.11 with an older `msbuild`?

**Comment by sipsorcery** [comment]
> May I ask you to build `Qt5.12.11_x64_static_vs2019_1694.zip` -- the recent Qt 5.12.11 with an older `msbuild`?

Sure, any idea where I can find the software versions installed on the cirrusci image you're using?

**Comment by hebasto** [comment]
> > May I ask you to build `Qt5.12.11_x64_static_vs2019_1694.zip` -- the recent Qt 5.12.11 with an older `msbuild`?
> 
> Sure, any idea where I can find the software versions installed on the cirrusci image you're using?

https://github.com/cirruslabs/docker-images-windows

**Comment by hebasto** [comment]
@sipsorcery 

The `msvc -version` output within the `cirrusci/windowsservercore:visualstudio2019` Cirrus image:

```
Microsoft (R) Build Engine version 16.9.0+5e4b48a27 for .NET Framework
Copyright (C) Microsoft Corporation. All rights reserved.
16.9.0.16703

**Comment by sipsorcery** [comment]
> The `msvc -version` output within the `cirrusci/windowsservercore:visualstudio2019` Cirrus image:
> 
> ```
> Microsoft (R) Build Engine version 16.9.0+5e4b48a27 for .NET Framework
> Copyright (C) Microsoft Corporation. All rights reserved.
> 16.9.0.16703
> ```

Thanks for doing that. I was trying to track it down via the docker files. I already have that version of msbuild installed so will attempt the Qt build tonight (it kills my machine so overnight job).

**Comment by hebasto** [comment]
@sipsorcery 

To make debugging a bit easier in the future, I've added
```
  build_environment_script:
    - choco list --localonly
    - msbuild -version
```

**Comment by sipsorcery** [comment]
@hebasto https://github.com/sipsorcery/qt_win_binary/releases/tag/qt51211x64_static_vs2019_160900

**Comment by hebasto** [comment]
> @hebasto https://github.com/sipsorcery/qt_win_binary/releases/tag/qt51211x64_static_vs2019_160900

Thanks!

**Comment by hebasto** [comment]
@MarcoFalke @fanquake @sipsorcery 

This PR is ready for reviewing now :tiger2:

**Comment by sipsorcery** [comment]
ACK ca9b9b9a2c0dd7ab767788be6512cfc34f1f6ee4.

Makes sense to consolidate CI platforms.

Tested that latest msbuild version `16.11.0.36601` still works with the `Qt5.12.11_x64_static_vs2019_160900` binaries.

One potential pitfall will be the `vcpkg` caching. We had [some issues](https://github.com/bitcoin/bitcoin/issues?q=is%3Aissue+vcpkg) with that in the past, namely whan a PR added or changed a dependency and the vcpkg couldn't deal with it. Since then the `vcpkg` manifest and binary caching features have been added. They may solve this problem.

**Comment by fanquake** [comment]
ACK after rebase

**Comment by hebasto** [comment]
Rebased ca9b9b9a2c0dd7ab767788be6512cfc34f1f6ee4 -> da8810e4f69b43e96ddf69400bec503451a0e847 ([pr21551.06](https://github.com/hebasto/bitcoin/commits/pr21551.06) -> [pr21551.07](https://github.com/hebasto/bitcoin/commits/pr21551.07)) due to the conflict with #22861.

File: .cirrus.yml [hunk 1]
```diff
@@ -89,6 +73,64 @@ task:
   env:
     << : *CIRRUS_EPHEMERAL_WORKER_TEMPLATE_ENV
 
+task:
+  name: "Win64 [unit tests, no functional tests] [msvc]"
+  windows_container:
+    cpu: 8
```
Comment (MarcoFalke) [inline .cirrus.yml:79]: why 8 when the average usage is 2?

File: .cirrus.yml [hunk 2]
```diff
@@ -89,6 +73,64 @@ task:
   env:
     << : *CIRRUS_EPHEMERAL_WORKER_TEMPLATE_ENV
 
+task:
+  name: "Win64 [unit tests, no functional tests] [msvc]"
```
Comment (MarcoFalke) [inline .cirrus.yml:77]:
```suggestion
  name: "Win64 native [unit tests, no functional tests] [msvc]"
```

File: .cirrus.yml [hunk 3]
```diff
@@ -89,6 +73,64 @@ task:
...
+  windows_container:
+    cpu: 8
+    memory: 7424Mi
+    image: cirrusci/windowsservercore:visualstudio2019
+  << : *CREDITS_TEMPLATE
```
Comment (MarcoFalke) [inline .cirrus.yml:82]: This looks expensive, especially with 8 CPUs and without `BASE_TEMPLATE` to skip

**Review by maflcko — COMMENTED** [review]
Missing BASE_TEMPLATE to merge the latest branch?

**Comment by hebasto** [comment]
Updated da8810e4f69b43e96ddf69400bec503451a0e847 -> 97292b19140db32c6d85d63b70382e7bf60a55c4 ([pr21551.07](https://github.com/hebasto/bitcoin/commits/pr21551.07) -> [pr21551.08](https://github.com/hebasto/bitcoin/commits/pr21551.08), [diff](https://github.com/hebasto/bitcoin/compare/pr21551.07..pr21551.08)):

- addressed @MarcoFalke's comments

File: .cirrus.yml (hunk 2 above)
Comment (hebasto) [reply .cirrus.yml]: Thanks! [Updated](https://github.com/bitcoin/bitcoin/pull/21551#issuecomment-913962298).

File: .cirrus.yml (hunk 3 above)
Comment (hebasto) [reply .cirrus.yml]: [Updated](https://github.com/bitcoin/bitcoin/pull/21551#issuecomment-913962298).

File: .cirrus.yml (hunk 1 above)
Comment (hebasto) [reply .cirrus.yml]: With `cpu: 4`:
 
Screenshot from 2021-09-07 06-45-03

And the total time with a populated `vcpkg` cache:

Screenshot from 2021-09-07 06-46-09

**Comment by MarcoFalke** [comment]
ack

**Comment by hebasto** [comment]
@sipsorcery 

Another (final?) review? :smile:

**Comment by sipsorcery** [comment]
re-ACK 97292b19140db32c6d85d63b70382e7bf60a55c4.

File: .cirrus.yml [hunk 4]
```diff
@@ -89,6 +76,73 @@ task:
...
+    VCPKG_FEATURE_FLAGS: 'manifests'
+    QT_DOWNLOAD_URL: 'https://github.com/sipsorcery/qt_win_binary/releases/download/qt51211x64_static_vs2019_160900/Qt5.12.11_x64_static_vs2019_160900.zip'
+    QT_DOWNLOAD_HASH: 'b24436bbc49ac69d992efc148e640f02e8dec426bed5f8497abf735e7d7d59d0'
+    QT_LOCAL_PATH: 'C:\Qt5.12.11_x64_static_vs2019_160900'
+    IgnoreWarnIntDirInTempDetected: 'true'
+  merge_script:
+    - git config --global user.email "ci@ci.ci"
+    - git config --global user.name "ci"
+    - git config core.filemode false
+    - git reset --hard
```
Comment (MarcoFalke) [inline .cirrus.yml:101]: maybe add a comment why this is needed?

File: .cirrus.yml [hunk 5]
```diff
@@ -89,6 +76,73 @@ task:
...
+  build_environment_script:
+    - choco list --localonly
+    - Write-Host ""
+    - msbuild -version
+  build_script:
+    - cd $env:CIRRUS_WORKING_DIR
+    - python build_msvc\msvc-autogen.py
+    - msbuild build_msvc\bitcoin.sln -property:Configuration=Release -maxCpuCount -verbosity:minimal -noLogo
+  unit_tests_script:
+    - src\test_bitcoin.exe
```
Comment (MarcoFalke) [inline .cirrus.yml:141]: why remove the `-l test_suite`?

Comment (hebasto) [reply .cirrus.yml:141]: :man_shrugging: Going to re-add right now.

File: .cirrus.yml (hunk 4 above)
Comment (hebasto) [reply .cirrus.yml:101]: Sure, I'm on it.

File: .cirrus.yml (hunk 5 above)
Comment (hebasto) [reply .cirrus.yml:141]: Done in #22909.

File: .cirrus.yml (hunk 4 above)
Comment (hebasto) [reply .cirrus.yml:101]: Done in #22909.

**Pull request merged**